overlord/snapstate/snapstate.go

```diff
@@ -529,13 +536,41 @@ func Update(st *state.State, name, channel string, revision snap.Revision, userI
 	return doInstall(st, &snapst, snapsup)
 }
 
+func validInfoForFlags(info *snap.Info, snapst *SnapState, flags Flags) bool {
+	switch c := info.Confinement; c {
+	case snap.StrictConfinement, "":
+		// strict is always fine
+		return true
+	case snap.DevModeConfinement:
+		// --devmode needs to be specified every time (==> ignore snapst)
+		return flags.DevModeAllowed()
+	case snap.ClassicConfinement:
+		if flags.Classic {
+			return true
+		}
+
+		if snapst != nil && snapst.Flags.Classic {
+			return true
+		}
+
+		return false
+	default:
+		logger.Panicf("unknown confinement %q", c)
```

Comment:
ok, i'll return `false` and log something. This'll get changed to returning `error` when I implement the errors themselves.